### ISSUE TITLE
[Request for External] Table 370 Excell Buffer; Codeunit 419 File Management

### ISSUE THREAD

**Issue by CKL-CAalstede** [microsoft/ALAppExtensions#3054]
I am having a tiny issue transfering our Reports that handle the Export to Excel into AL Language. All issues are created because certain Functions that are set to Global in C/AL Coding have not yet been made External for Extension development.
To fully achieve our conversion to AL we would need the following Functions set to External:

Table 370 Excel Buffer - Select Sheets Name
**_[External]_** SelectSheetsName(FileName : Text) : Text[250]
IF FileName = '' THEN
ERROR(Text001);
FileManagement.BLOBImportFromServerFile(TempBlob,FileName);
TempBlob.Blob.CREATEINSTREAM(InStr);
EXIT(SelectSheetsNameStream(InStr));

Table 370 Excel Buffer - Update Book
**_[External]_** UpdateBook(FileName : Text;SheetName : Text)
UpdateBookExcel(FileName,SheetName,TRUE);

Codeunit 419 File Management - Upload File
_**[External]**_ UploadFile(WindowTitle : Text[50];ClientFileName : Text) ServerFileName : Text
Filter := GetToFilterText('',ClientFileName);
IF PathHelper.GetFileNameWithoutExtension(ClientFileName) = '' THEN
ClientFileName := '';
ServerFileName := UploadFileWithFilter(WindowTitle,ClientFileName,Filter,AllFilesFilterTxt);

**Comment by RomanZanizdra** [comment]
Hi. These functions cannot be made external and available for AL development. I suggest you to use functions with almost the same names (they contain Stream word) for your purposes. For example "SelectSheetsNameStream". 
For "UploadFile" - you should find another function fitting your requirements.

**Comment by CKL-CAalstede** [comment]
Hey, I could succesfully change our logic to base ourselves on the Stream functions you mentioned for the UpdateBook - UpdateBookStream and SelectSheetName - SelectSheetNameStream.
Selecting another function that imitates UploadFile has however been quite a challenge. Right now I am making use of the BLOBImport (Codeunit 419) function but this comes with some restrains.

1. The Filename is lost. I can't figure out where the File Name of the to be updated Excel File is saved in this function and as such, the File Name goes lost when trying to save the Updated Version.

2. The Upload prompt is differs to the Business Central Standard "Update Existing Worksheet" prompt. This BLOBImport function simply displays Import, without me being able to change this prompt in any way.

Can you in any way help me figure these 2 issues out. From my perspective these 2 are big burdens on the User Experience and we strive to deliver the best solutions!

Kind Regards,
Ciprian van Aalstede

**Comment by RomanZanizdra** [comment]
Hi Ciprian. Can you describe your business case? What do you need to get? I don't fully understand the scenario.

**Comment by CKL-CAalstede** [comment]
Hey Roman,

when exporting our own Schedule Matrix, that analyzes Product Specific Entries, to Excel we tried our best to keep our Code the same with Standard Coding. As such we use the Standard Excel Buffer Table and it's functionality, as well as the UploadFile Function from the FileManagement Codeunit 419.

The issue arised because all usually Global Functions of the Excel buffer and FileManagement Codeunit were not open for External use. The issue only arrises for the "Update Existing Worksheet" functionality.

As you proposed I changed our Coding from using the String Functions of the Excel Buffer Table to the Stream Word Functions, sucessfuly. 

Right now the only issue is that the only function I could find that is similar to the UploadFile Function, and that is open for External use, is BLOBImport Function.

Right now the Export process with the option "Update Existing Worksheet" is functional, however there are massive changes in the User Experience. As per my comment Above the following issues arise:
1. The Filename that the user Selects to be Updated is lost. I can't figure out where the File Name of the to be updated Excel File is saved in this function and as such, the File Name goes lost when trying to save the Updated Version.
2. The Upload prompt is different to the Business Central Standard "Update Existing Worksheet" prompt. This BLOBImport function simply displays "Import" instead of the known "Update Workbook", without me being able to change this prompt in any way.

**Comment by RomanZanizdra** [comment]
Sorry, but still not clear.
1) What is the "Update Existing Worksheet" functionality? Where is it? Is this standard functionality or yours?
2) What is the "Upload prompt" and "Business Central Standard "Update Existing Worksheet" prompt"? Is this also standard functionality?
3) One more time about scenario. Can you describe it the following way: user exports data into the excel file, then he edits it, then he uploads it to blob or ….. something else?
I understand the root cause of your problem, because now you cannot use some functions which was available for development in OnPrem. But I need information I sked in order I can suggest you something valuable or at least clearly understand the problem and try to find out with you the solution.

**Comment by CKL-CAalstede** [comment]
No issue :), you are helping me out so I'm sorry for not clarifying it more precisely.
1. "Update Existing Worksheet" is a Standard functionality that can be found, for example in the Account Schedules, in the Matrix Overview(under Process), under Actions - Excel - Export to Excel - Update Existing Document(seems I messed up the name, I apologise). 
This is the  same functionality we used to have for the Export to Excel - Update Existing Document of our own ISV Specific Schedules.(Only true for OnPrem Version now)

2. Because the "Update Existing Document" is no longer available I searched for a good replacement function and ran into the BLOBImport function. This is a Standard function as well and can be found in Codeunit 419. Sadly I cannot say for what functionality in Business Central Standard it is used as I have never run into it before. The are two big differences between the two:
                            i. The File Name is not saved in the BLOBImport function.
                            ii. The Message that asks the User to select a File differs.

3. The user tries to Export data into an already exising Excel File. He is first of all asked to which existing File he wants to write, at this Point the message that is shown in our "Update Existing Document" differs to the on that is shown in the Business Central Standard "Update Existing Document" due to the use of different functions (BLOBImport/UploadFile). The uploading of the Excel File to Blob function is so that the File can be Uploaded from the Client Machine to the Server Machine where Data from the Schedule is written in the selected Worksheet. The User is then asked to save the file where the second issue arises. The File Name, of the File he selected to be updated, is no longer recognized by Business Central because it is not saved in within the BLOBImport function, unlike the UploadFile function where it is saved.

I hope this makes my issue a bit clearer :).
